nexus/src/db/datastore/project.rs

```diff
@@ -154,28 +185,48 @@ impl DataStore {
 
         use db::schema::project::dsl;
 
-        let now = Utc::now();
-        let updated_rows = diesel::update(dsl::project)
-            .filter(dsl::time_deleted.is_null())
-            .filter(dsl::id.eq(authz_project.id()))
-            .filter(dsl::rcgen.eq(db_project.rcgen))
-            .set(dsl::time_deleted.eq(now))
-            .returning(Project::as_returning())
-            .execute_async(self.pool_authorized(opctx).await?)
-            .await
-            .map_err(|e| {
-                public_error_from_diesel_pool(
-                    e,
-                    ErrorHandler::NotFoundByResource(authz_project),
+        type TxnError = TransactionError<Error>;
+        self.pool_authorized(opctx)
+            .await?
+            .transaction_async(|conn| async move {
+                let now = Utc::now();
+                let updated_rows = diesel::update(dsl::project)
+                    .filter(dsl::time_deleted.is_null())
+                    .filter(dsl::id.eq(authz_project.id()))
+                    .filter(dsl::rcgen.eq(db_project.rcgen))
+                    .set(dsl::time_deleted.eq(now))
+                    .returning(Project::as_returning())
+                    .execute_async(&conn)
+                    .await
+                    .map_err(|e| {
+                        public_error_from_diesel_pool(
+                            PoolError::from(e),
+                            ErrorHandler::NotFoundByResource(authz_project),
+                        )
+                    })?;
+
+                if updated_rows == 0 {
+                    return Err(TxnError::CustomError(Error::InvalidRequest {
+                        message:
+                            "deletion failed due to concurrent modification"
```

Comment:
Again, it looks like this is not changed from before, but I think this should be a 500 error.  I'd also strongly recommend reporting the fact that we know, which is that we unexpectedly updated zero rows, rather than our inference that there was a concurrent modification.

Reply:
https://github.com/oxidecomputer/omicron/search?q=deletion+failed+due+to+concurrent+modification

I'm pretty sure this is the same error (and message) used elsewhere in omicron; are you okay punting this to a different PR?